states/cacheimports.go

```diff
@@ -4,35 +4,35 @@ import "sync"
 
 // CacheImports is a synchronized set of cache imports.
 type CacheImports struct {
-	mu    sync.Mutex
-	store map[string]bool
+	mu    sync.RWMutex
+	store []string
 }
 
 // NewCacheImports creates a new cache imports structure.
-func NewCacheImports(imports map[string]bool) *CacheImports {
-	store := make(map[string]bool)
-	for k, v := range imports {
-		store[k] = v
-	}
+func NewCacheImports(imports []string) *CacheImports {
+	clone := make([]string, len(imports))
+	copy(clone, imports)
+
 	return &CacheImports{
-		store: store,
+		store: clone,
 	}
 }
 
-// Add adds an import to the set.
-func (ci *CacheImports) Add(tag string) {
+// Add adds imports to the set.
+func (ci *CacheImports) Add(tags ...string) {
 	ci.mu.Lock()
 	defer ci.mu.Unlock()
-	ci.store[tag] = true
```

Comment:
I'm worried about loosing this functionality. Can we keep both the map and list, effectively turning this into an ordered set?

Reply:
@alexcb 
Thank you for your comment!
OK. I will add a map into `CacheImports`.
Should I implement `AsMap() map[string]bool` or `Has(string tag) bool` in `CacheImports`?

Reply:
I think it's fine to simply keep `AsSlice()`; I don't think it's possible to implement `AsMap()` given that ordering is now important.

`Has(...)` sounds like a useful method.

Reply:
@alexcb 
FWI, I can also look up a specific tag from the slice of `CacheImports` in `Has()` method.
Should I add a map into `CacheImport` or look up a tag from the slice?

Reply:
@alexcb 
I implemented `Has()` method with a map for now.
[7671cd0](https://github.com/earthly/earthly/pull/2063/commits/7671cd029174a40d43d976d61043ef5053a27d63)

Reply:
The logic here seems wrong. we only want to Add the tag if it doesn't already exist in the set:

```
if _, exists := ci.store[tag]; exists {
    return
}
ci.store[tag] = true
ci.slice = append(.......
```

Reply:
Fix by [66decf1](https://github.com/earthly/earthly/pull/2063/commits/66decf19254b23661cb11e2aafc5da402c444af2)